### PR TITLE
BME280 - Humidity Support

### LINKED ISSUE #399: Add Hummidity and Compensate Hummidity to the BME280 device binding.
For BME280, you can read the pressure and the temperature (including compensate values) but so far there's nothing for humidity. Any chance for this to be implemented please?

### PR DESCRIPTION
**Pull request by Silvenga**
This pr adds support for humidity for BME280 devices.

- Moved the responsibility of reading calibration data to the implemented devices, away from `CalibrationData`.
- Added additional registers.
- Added method to read humidity data and to apply calibrations.

This resolves #399.

### CODE REVIEW

**Comment by Silvenga** [comment]
Hopefully did everything correctly. First time reading data sheets and messing with iot sensors. Plus it turns out to be almost midnight here. 😸

**Comment by Silvenga** [comment]
And whoa, that code was really bad. Re-read the data sheets and fixed the bugs. Compared against the python version for correctness.

File: src/devices/Bmx280/samples/Bme280.sample.cs [hunk 1]
```diff
@@ -23,53 +23,70 @@ static async Task Main(string[] args)
             //set this to the current sea level pressure in the area for correct altitude readings
             const double defaultSeaLevelPressure = 1033.00;
 
-            var i2cSettings = new I2cConnectionSettings(busId, Bme280.DefaultI2cAddress);
+            var i2cSettings = new I2cConnectionSettings(busId, Bme280.DefaultI2cAddress + 1);
```
Comment (joperezr) [inline src/devices/Bmx280/samples/Bme280.sample.cs]: I did also find that the default I2C Address is incorrect. Do you mind changing the constant on the same PR to use 0x77 instead?

File: src/devices/Bmx280/samples/Bme280.sample.cs [hunk 2]
```diff
@@ -23,53 +23,70 @@ static async Task Main(string[] args)
...
             var i2cDevice = new UnixI2cDevice(i2cSettings);
             var i2CBmpe80 = new Bme280(i2cDevice);
 
             using (i2CBmpe80)
             {
                 while (true)
                 {
-                    ////set mode forced so device sleeps after read
+                    Console.WriteLine("Let's get this party started!");
```
Comment (joperezr) [inline src/devices/Bmx280/samples/Bme280.sample.cs]: NIT: Nice! but we should probably remove it :)

File: src/devices/Bmx280/Bmx280.csproj [hunk 3]
```diff
@@ -3,16 +3,19 @@
   <PropertyGroup>
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <EnableDefaultItems>false</EnableDefaultItems>
+    <RootNamespace>Iot.Device.Bmx280</RootNamespace>
```
Comment (joperezr) [inline src/devices/Bmx280/Bmx280.csproj]: Why did you need to add this?

File: src/devices/Bmx280/BmxBase.cs [hunk 4]
```diff
@@ -13,13 +13,14 @@
...
-    public class BmxBase : IDisposable
+    public abstract class BmxBase : IDisposable
     {
         internal I2cDevice _i2cDevice;
         internal byte _deviceId;
         internal bool _initialized = false;
         internal CommunicationProtocol _communicationProtocol;
-        internal CalibrationData _calibrationData;
+
+        internal CalibrationData CalibrationData { get; private set; }
```
Comment (joperezr) [inline src/devices/Bmx280/BmxBase.cs]: Why do we want to change it to a property instead of keeping it as a member of the class?

File: src/devices/Bmx280/BmxBase.cs [hunk 5]
```diff
@@ -305,24 +307,87 @@ private long CompensatePressure(int adcPressure)
...
+        /// <summary>
+        /// Get the current status of the device.
+        /// </summary>
+        /// <returns></returns>
+        public DeviceStatus ReadStatus()
+        {
+            var status = Read8BitsFromRegister((byte)Register.STATUS);
+
+            // Bit 3
+            var measuring = (status & 0x0000_1000) >> 3 == 1;
```
Comment (joperezr) [inline src/devices/Bmx280/BmxBase.cs]: NIT: You could do the shift of status first and then just make & 1 check.

File: src/devices/Bmx280/BmxBase.cs [hunk 6]
```diff
@@ -305,24 +307,87 @@ private long CompensatePressure(int adcPressure)
...
+
+            return new DeviceStatus
+            {
+                ImageUpdating = imUpdate,
+                Measuring = measuring
+            };
+        }
+
+        /// <summary>
+        /// Sets the IIR filter mode.
```
Comment (joperezr) [inline src/devices/Bmx280/BmxBase.cs]: Do you mind explaining what IIR means?

File: src/devices/Bmx280/Register.cs [hunk 7]
```diff
@@ -23,14 +23,25 @@ internal enum Register : byte
...
+        RESET = 0xE0,
 
         CAL26 = 0xE1,  // R calibration stored in 0xE1-0xF0
 
+        // BME280 Only
+        CTRL_HUM = 0xF2,
+
         STATUS = 0xF3,
-        CONTROL = 0xF4,
+        CTRL_MEAS = 0xF4,
```
Comment (joperezr) [inline src/devices/Bmx280/Register.cs]: I suppose this is the name in the DataSheet?

File: src/devices/Bmx280/DeviceStatus.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Iot.Device.Bmx280
+{
+    public class DeviceStatus
```
Comment (joperezr) [inline src/devices/Bmx280/DeviceStatus.cs]: Do you mind adding doc comments to this class and to the FilteringMode as well?

**Comment by joperezr** [comment]
Left a few comments but overall this looks great! Thanks so much for adding this @Silvenga

File: src/devices/Bmx280/samples/Bme280.sample.cs (hunk 1 above)
Comment (Silvenga) [reply src/devices/Bmx280/samples/Bme280.sample.cs]: Oops, didn't mean to commit that. I'll update the constant.

File: src/devices/Bmx280/samples/Bme280.sample.cs (hunk 2 above)
Comment (Silvenga) [reply src/devices/Bmx280/samples/Bme280.sample.cs]: I did that mainly so I could see what loop we were on (I was using the sample app to test with). I'll pull it out.

File: src/devices/Bmx280/Bmx280.csproj (hunk 3 above)
Comment (Silvenga) [reply src/devices/Bmx280/Bmx280.csproj]: That was to ensure that newly created/refactored files used the correct namespace. Otherwise the default, defaults to just `Bmx280` IIRC. This also makes static analysis happy (Resharper really wanted to fix that namespace).

Comment (Silvenga) [reply src/devices/Bmx280/Bmx280.csproj]: Should I pull it out?

File: src/devices/Bmx280/BmxBase.cs (hunk 4 above)
Comment (Silvenga) [reply src/devices/Bmx280/BmxBase.cs]: I wanted `set` restrictions on this field (hence the property refactor). 

The base class owns `_calibrationData`, but implementations need read access too. Since the `Init` operation that creates  `_calibrationData` is both owned by `BmxBase` and is lazily initialized by it, it made sense to me to refactor `_calibrationData` to be a property on `BmxBase`.

File: src/devices/Bmx280/Register.cs (hunk 7 above)
Comment (Silvenga) [reply src/devices/Bmx280/Register.cs]: Yep.

File: src/devices/Bmx280/BmxBase.cs (hunk 6 above)
Comment (Silvenga) [reply src/devices/Bmx280/BmxBase.cs]: It's what the datasheet uses to refer to the internal logic that reduces noise at the cost of accuracy - Infinite impulse response (IIR). The datasheet talks about different recommended filter settings based on usage. The IIR circuits are only used for temperature and and pressure.

File: src/devices/Bmx280/DeviceStatus.cs (hunk 8 above)
Comment (Silvenga) [reply src/devices/Bmx280/DeviceStatus.cs]: Will do.

**Comment by Silvenga** [comment]
Thanks @joperezr - also had a question about the timing intervals. I wanted to add support, but I have no idea how to name the enum members. This config flag controls the detection/sleep interval when in `normal` mode. The values are kind of odd to build into an enum. I wanted to use an abstraction like the power saving enum for over sampling, but can't find a compatible name.
```
0.5 (ReallyFast)
62.5ms (KindOfFast)
125ms (LessFastThenKindOfFastButSlowerThanFast)
250ms (Fast)
...
10ms
20ms
```

I was think we could alternatively support a dictionary of possible values (with the mapping to their byte version), and have the user pass in a ms that they want to use, but that's not exactly intuitive.

File: src/devices/Bmx280/Bmx280.csproj (hunk 3 above)
Comment (joperezr) [reply src/devices/Bmx280/Bmx280.csproj]: Oh ok, no it's fine to leave it. I was just wondering so that if it was useful we would add it to the binding template. Sounds like it may be good to add it there at some point.

File: src/devices/Bmx280/BmxBase.cs (hunk 4 above)
Comment (joperezr) [reply src/devices/Bmx280/BmxBase.cs]: oh ok, that makes sense.

**Comment by joperezr** [comment]
Regarding your timing intervals question, does the DataSheet provide a naming for the different intervals? If not, what about having the enum literally tell what the delay is? Another option would be to have this be a TimeSpan property and when set, we would simply kind of round to the closest one that the hardware allows, although I'm not a fan of that approach.

**Comment by Silvenga** [comment]
@joperezr I think I got everything. I added support for standby times - the enum is a little odd. I also added  the licensed header that was missing from a couple of files.

File: src/devices/Bmx280/samples/Bme280.sample.cs [hunk 9]
```diff
@@ -23,53 +23,68 @@ static async Task Main(string[] args)
             //set this to the current sea level pressure in the area for correct altitude readings
             const double defaultSeaLevelPressure = 1033.00;
 
-            var i2cSettings = new I2cConnectionSettings(busId, Bme280.DefaultI2cAddress);
+            var i2cSettings = new I2cConnectionSettings(busId, Bme280.DefaultI2cAddress + 1);
```
Comment (joperezr) [inline src/devices/Bmx280/samples/Bme280.sample.cs]: I suppose you don't need this +1 any longer now that you've fixed the address?

**Comment by joperezr** [comment]
Thanks for adding this @Silvenga ! I will go ahead and merge as soon as you remove that +1 from the sample as I have no further comments.

File: src/devices/Bmx280/samples/Bme280.sample.cs (hunk 9 above)
Comment (Silvenga) [reply src/devices/Bmx280/samples/Bme280.sample.cs]: Oops, fixed!

**Comment by Silvenga** [comment]
@joperezr I rebased onto master. Anything preventing a merge at this point?

**Comment by joperezr** [comment]
I'm so sorry about not merging this earlier, I lost track of the PR since I had already signed off on it. We are actually in the middle of a somewhat big refactoring for this binding here: #477. Is there any chance that we can incorporate your changes into that PR since most of the files that you change here are getting moved? cc: @georgemathieson It would be great if we could add this code to your PR so that Bme280 would also have Humidity Support.

**Comment by Silvenga** [comment]
Not a problem, I can rebase onto georgemathieson:bme680 if that's where we are headed. Just let me know what to do.

**Comment by joperezr** [comment]
That would be awesome! I'm very sorry again for not merging this earlier so that it got picked up by his refactoring 😨. @georgemathieson I suppose you are ok with @Silvenga pushing into your PR branch? Or do you prefer to merge yours first and have him create a new one on top of that?

**Comment by georgemathieson** [comment]
Hey @joperezr I think if this branch went into master then I would be happy to bring master into my branch, and bring in the functionality that way. What are your thoughts? I'm just conscious that #482 and @RobinTTY are also blocked by the refactor.

**Comment by joperezr** [comment]
@georgemathieson That sounds good to me. In that case, I'll go ahead and merge this PR. Thanks a lot for understanding and for your refactoring work!
